lib/websocket/connection.js

```diff
@@ -0,0 +1,409 @@
+
+      receiveData(ws)
+
+      socketClosed(ws)
+    }
+  })
+
+  return controller
+}
+
+/**
+ * @param {import('../fetch/index').Fetch} controller
+ * @param {import('net').Socket|import('tls').TLSSocket|undefined} socket
+ */
+function failWebsocketConnection (controller, socket) {
+  controller.abort()
+
+  if (socket && !socket.destroyed) {
+    socket.destroy()
+  }
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function whenConnectionEstablished (ws) {
+  const { [kResponse]: response } = ws
+
+  // 1. Change the ready state to OPEN (1).
+  ws[kReadyState] = states.OPEN
+
+  // 2. Change the extensions attribute’s value to the extensions in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-9.1
+  const extensions = response.headersList.get('sec-websocket-extensions')
+
+  if (extensions !== null) {
+    ws[kExtensions] = extensions
+  }
+
+  // 3. Change the protocol attribute’s value to the subprotocol in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-1.9
+  const protocol = response.headersList.get('sec-websocket-protocol')
+
+  if (protocol !== null) {
+    ws[kProtocol] = protocol
+  }
+
+  // 4. Fire an event named open at the WebSocket object.
+  fireEvent('open', ws)
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function receiveData (ws) {
+  const { [kResponse]: response } = ws
+
+  /** @type {WebsocketFrame|undefined} */
+  let frame
+
+  response.socket.on('data', (chunk) => {
+    const receivedFrame = WebsocketFrame.from(chunk)
+
+    if (!frame) {
+      frame = receivedFrame
+      frame.terminated = frame.fin // message complete
+    } else {
+      // A fragmented message consists of a single frame with the FIN bit
+      // clear and an opcode other than 0, followed by zero or more frames
+      // with the FIN bit clear and the opcode set to 0, and terminated by
+      // a single frame with the FIN bit set and an opcode of 0.
+
+      if (receivedFrame.fin && receivedFrame.opcode === opcodes.CONTINUATION) {
+        frame.terminated = true
+      }
+
+      frame.addFrame(receivedFrame.data)
+    }
+
+    // If a control frame (Section 5.5) is
+    // received, the frame MUST be handled as defined by Section 5.5.
+    if (frame.opcode === opcodes.PING) {
+      handlePing(response.socket, ws)
+
+      return
+    } else if (frame.opcode === opcodes.PONG) {
+      handlePong(frame)
+
+      return
+    } else if (frame.opcode === opcodes.CLOSE) {
+      // Upon either sending or receiving a Close control frame, it is said
+      // that _The WebSocket Closing Handshake is Started_ and that the
+      // WebSocket connection is in the CLOSING state.
+      ws[kReadyState] = states.CLOSING
+      ws[kClosingFrame] = frame
```

Comment:
When you receive a close frame, you should send back another close frame. The connection should be closed only when both peers have sent and received a close frame.

Reply:
Thanks! This actually fixed an issue where the wpt runner would take ~35 seconds to run (I'm assuming the server was staying open because it wasn't receiving a close frame). Now it takes 4 seconds to run.